### DISCUSSION TITLE
How do I Bind a Collection in Conjunction with WhenActivated Code-Behind

### DISCUSSION THREAD

**Discussion by FaithBeam** [AvaloniaUI/Avalonia#17736] [Q&A] [answered]
Hello! I've been pulling my hair out with this problem. I'd like to use WhenActivated in the view model and view's code-behind and bind a collection to a combobox's itemssource. I cannot get this to work when I use WhenActivated in the view's code-behind, but it works when I use bindings from xaml.

Full reproducible project here: https://github.com/FaithBeam/AvaloniaWhenActivatedCollectionBinding

MainWindowViewModel.cs:
```csharp
using System;
using System.Collections.ObjectModel;
using System.Reactive.Disposables;
using DynamicData;
using ReactiveUI;

namespace AvaloniaWhenActivatedCollectionBinding.ViewModels;

public partial class MainWindowViewModel : ViewModelBase, IActivatableViewModel
{
    private readonly ReadOnlyObservableCollection<string> _strings;
    public MainWindowViewModel()
    {
        Activator = new ViewModelActivator();
        var sourceList = new SourceList<string>();
        sourceList.Connect().Bind(out _strings).Subscribe();
        this.WhenActivated(d =>
        {
            sourceList.Edit(x => x.Add("First"));
            this.RaisePropertyChanged(nameof(Strings));
            Disposable.Create(() => { }).DisposeWith(d);
        });
    }
    public ReadOnlyObservableCollection<string> Strings => _strings;
    public ViewModelActivator Activator { get; }
}
```

MainWindow.xaml.cs:
```csharp
using System.Reactive.Disposables;
using Avalonia.Controls;
using Avalonia.ReactiveUI;
using AvaloniaWhenActivatedCollectionBinding.ViewModels;
using ReactiveUI;

namespace AvaloniaWhenActivatedCollectionBinding.Views;

public partial class MainWindow : ReactiveWindow<MainWindowViewModel>
{
    public MainWindow()
    {
        InitializeComponent();
        this.WhenActivated(d =>
        {
            this.OneWayBind(ViewModel, vm => vm.Strings, v => v.ComboBox.ItemsSource).DisposeWith(d);
        });
    }
}
```

The combobox has an item, but there's no text...
![code-behind-binding](https://github.com/user-attachments/assets/3be43060-0525-44c9-9c27-cd3ce771fc96)

And once again, when I comment out the code-behind's when activated block and bind the combobox's items source in xaml, it works, but I'd like to use the code-behind :).

Xaml binding works for reference:
![xaml-binding](https://github.com/user-attachments/assets/0e5471c4-5f62-49be-908a-bf3615c442c2)

Any ideas? Thank you!

**Comment by thevortexcloud** [comment]
Does it work if you use a dispatcher call? From memory ReactiveUI uses its own scheduler thread. But this may be a different problem entirely.

**Reply by FaithBeam** [reply]
Hello! Do you mean this?
```csharp
using System;
using System.Collections.ObjectModel;
using System.Reactive.Disposables;
using Avalonia.Threading;
using DynamicData;
using ReactiveUI;

namespace AvaloniaWhenActivatedCollectionBinding.ViewModels;

public partial class MainWindowViewModel : ViewModelBase, IActivatableViewModel
{
    private readonly ReadOnlyObservableCollection<string> _strings;

    public MainWindowViewModel()
    {
        Activator = new ViewModelActivator();
        var sourceList = new SourceList<string>();
        sourceList.Connect().Bind(out _strings).Subscribe();
        this.WhenActivated(d =>
        {
            Dispatcher.UIThread.Invoke(() =>
            {
                sourceList.Edit(x => x.Add("First"));
            });

            this.RaisePropertyChanged(nameof(Strings));
            Disposable.Create(() => { }).DisposeWith(d);
        });
    }

    public ReadOnlyObservableCollection<string> Strings => _strings;
    public ViewModelActivator Activator { get; }
}
```

If this is what you're talking about, I still get the same result as before.
![image](https://github.com/user-attachments/assets/f7285c60-d560-4f58-827c-4f08cb1c1d97)

**Reply by IanRawley** [reply, marked as answer]
I had a look and it is binding properly, but the ComboBox is failing to find a DataTemplate for the new child and so not presenting anything.  If you manually set a DataTemplate like below everything shows up fine.  I'm not sure why it's not falling back on the default template.
```
<ComboBox.ItemTemplate>
	<DataTemplate>
		<TextBlock Text="{Binding}"/>
	</DataTemplate>
</ComboBox.ItemTemplate>
```

There are multiple errors logged without the template.
DefaultViewLocator: Failed to resolve view for view model type 'System.Object'.
ViewModelViewHost: Couldn't find view for 'First'. Is it registered? Falling back to default content.

More Digging:
Avalonia implements a BindingHook that auto assigns the ItemTemplate of ItemsControls to a ViewModelViewHost if not otherwise set when you use the reactive Bind mixins.  When you do the binding with BindOneWay without an ItemTemplate, Avalonia automatically assigns a ViewModelViewHost which can't resolve a template for a String.  If you do the bind in XAML the ItemTemplate stays empty, but goes through the regular Avalonia methods and so finds a default TextBlock template.

**Reply by FaithBeam** [reply]
That's the ticket! Thank you very much for investigating this.
